src/jailer/src/env.rs

```diff
@@ -197,37 +200,53 @@ impl Env {
     }
 
     fn exec_into_new_pid_ns(&mut self, chroot_exec_file: PathBuf) -> Result<()> {
-        // Unshare into a new PID namespace.
-        // The current process will not be moved into the newly created namespace, but its first
-        // child will assume the role of init(1) in the new namespace.
-        // The call is safe because we're invoking a C library function with valid parameters.
-        SyscallReturnCode(unsafe { libc::unshare(libc::CLONE_NEWPID) })
-            .into_empty_result()
-            .map_err(Error::UnshareNewPID)?;
-
         // Compute jailer's total CPU time up to the current time.
         self.jailer_cpu_time_us =
             utils::time::get_time_us(utils::time::ClockType::ProcessCpu) - self.start_time_cpu_us;
 
-        // Duplicate the current process. The child process will belong to the previously created
-        // PID namespace.
-        // TODO: replace the `unshare()` + `fork()` combo with `clone()` if we ever need to
-        //  squeeze every bit of start-up latency we can get
-        let pid = unsafe { libc::fork() };
-        match pid {
-            0 => {
-                // Reset process start time.
-                self.start_time_cpu_us = 0;
-
-                Err(Error::Exec(self.exec_command(chroot_exec_file)))
-            }
-            child_pid => {
-                // Save the PID of the process running the exec file provided
-                // inside <chroot_exec_file>.pid file.
-                self.save_exec_file_pid(child_pid, chroot_exec_file)?;
-                unsafe { libc::exit(0) }
-            }
+        // This is a wrapper for our child callback to the clone libc function.
+        // We will use clone's arg parameter as the actual callback and call it from here.
+        extern "C" fn child_fn(arg: *mut libc::c_void) -> libc::c_int {
+            let callback_ref: *mut *mut dyn FnOnce() = arg as *mut *mut dyn std::ops::FnOnce();
+            let callback = unsafe { Box::from_raw(*callback_ref) };
+
+            callback();
+            0
         }
+
+        let child_callback = || {
+            // Reset process start time.
+            self.start_time_cpu_us = 0;
+
+            self.exec_command(chroot_exec_file.clone());
+        };
+
+        let child_callback_box: Box<dyn FnOnce()> = Box::new(child_callback);
+        let child_callback_ref = &mut Box::into_raw(child_callback_box);
+
+        // Mapping the child process it's own stack.
+        let stack: &mut [u8; PROCESS_STACK_SIZE] = &mut [0; PROCESS_STACK_SIZE];
+
+        // The call is safe because we're adding the stack size to itself for using it as stack.
+        let stack_ptr = unsafe { stack.as_mut_ptr().add(stack.len()) };
```

Comment:
> It's not very clear to me what will happen next from rust's perspective. Are we sure that rust is not going to clear this memory when the parent process executes unsafe { libc::exit(0) } ? Would we need to use mem::forget() here to make sure ? Regardless, please also add a comment that explains this in more detail.

I've had this same concern. But after the `clone` call returns, the entire memory will have already been cloned to the new process (since it's not using `CLONE_VM `). So from the new process perspective, rust may do anything to the parent.

Reply:
@serban300 - It will not work. The libc clone wrapper forbids passing a NULL as stack (even if not using CLONE_VM flag). We can call directly to the syscall - but this will be super ugly in my opinion.

Reply:
Ok, I see. Then I'm ok with providing the stack.